### PR TITLE
Add basic snippets to reset password

### PR DESCRIPTION
**Pull request by bjarnef**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description
Besides the general Login snippets, I think it would be useful to include common functionality to request password reset and reset the password. The current snippets include to basic to register, login, edit profile, but not an option to change password.

We have the following controllers `UmbLoginStatusController` and `UmbLoginController`.

The `UmbLoginController` has a method to handle login and `UmbLoginStatusController` has a method to handle log out.

Maybe we should move password functionality to a separate controller instead of the adding these to the existing controllers?

For backoffice users we have this to send invite email https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Web.BackOffice/Controllers/UsersController.cs#L568 and this to change password. https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Web.BackOffice/Controllers/UsersController.cs#L721

### CODE REVIEW

**Comment by emmagarland** [comment]
Hi @bjarnef 

Noticed this one in the list of PRs. I know it is in draft, but (other than the conflicts), is there anything we can do to help you progress this PR to a ready state? Did you want us to check regarding which controller to add the password functionality to?

Also, since this PR was created the new links for your examples are:

- Backoffice user invite email: https://github.com/umbraco/Umbraco-CMS/blob/0a4a938b887bac7f2c4a74715fda2c7780b03c03/src/Umbraco.Web.BackOffice/Controllers/UsersController.cs#LL569C4-L569C4
- Backoffice change password: https://github.com/umbraco/Umbraco-CMS/blob/0a4a938b887bac7f2c4a74715fda2c7780b03c03/src/Umbraco.Web.BackOffice/Controllers/UsersController.cs#L732

Thanks, 

Emma

**Comment by bjarnef** [comment]
Hi @emmagarland 

Yes, not sure if we want to add this to existing controllers or we should add a new.
Also also please check the reset password link, but I recall it was similar to the link sent in invite email.

I will look at the conflicting changes as well.

**Comment by bjarnef** [comment]
Hi @emmagarland

I have resolved the conflicts.
While working on this I noticed macros don't render in the legacy grid layout editor: https://github.com/umbraco/Umbraco-CMS/issues/14587

I have moved logic to a similar private method as in `SendUserInviteEmailAsync()`.

However from the generated token we can't use this https://github.com/umbraco/Umbraco-CMS/blob/a936fe09ab9c9c36130d5f8468dd21143ca39dc5/src/Umbraco.Web.BackOffice/Controllers/BackOfficeController.cs#L148 as with users. How should it works for members?

We can probably use a similar approach using language keys from member forgotten password mail as developers can translate/override these, but should probably just use culture from context/request?
https://github.com/umbraco/Umbraco-CMS/blob/0a4a938b887bac7f2c4a74715fda2c7780b03c03/src/Umbraco.Web.BackOffice/Controllers/UsersController.cs#L604-L610

**Comment by emmagarland** [comment]
Hi @bjarnef 

Thanks for this! It does seem a beneficial PR to have as an example for people to follow.

I have asked on the HQ channel re your `VerifyInvite` question for members. 

Best wishes

Emma

File: src/Umbraco.Web.Website/Controllers/UmbLoginController.cs [hunk 1]
```diff
@@ -33,12 +51,133 @@ public class UmbLoginController : SurfaceController
...
+        if (!_emailSender.CanSendRequiredEmail())
+        {
+            return CurrentUmbracoPage();
+        }
+
+        MemberIdentityUser? memberIdentity = await _memberManager.FindByNameAsync(model.Username);
+        if (memberIdentity == null!)
+        {
+            return new ValidationErrorResult(
+                $"No local member found for username {model.Username}");
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.Website/Controllers/UmbLoginController.cs:120]: This is not a good security practice, you're giving a potential attacker information about which email addresses are / are not registered on this site. The best practice is to return a message like: "If this is a known account then you should receive instructions in your email shortly."

File: src/Umbraco.Web.Website/Controllers/UmbLoginController.cs [hunk 2]
```diff
@@ -33,12 +51,133 @@ public class UmbLoginController : SurfaceController
...
+        }
+
+        // send the email
+        await SendMemberResetEmailAsync(memberIdentity, null, null, null);
+
+        TempData["ForgottenPasswordSuccess"] = true;
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.Website/Controllers/UmbLoginController.cs:126]: This isn't used anywhere, also as noted we should indicate neither success nor failure after submitting the form so as not to reveal if the account exists.

File: src/Umbraco.Web.Website/Controllers/UmbLoginController.cs [hunk 3]
```diff
@@ -33,12 +51,133 @@ public class UmbLoginController : SurfaceController
...
+        if (string.IsNullOrWhiteSpace(token) || string.IsNullOrWhiteSpace(model.UserId))
+        {
+            return CurrentUmbracoPage();
+        }
+
+        MemberIdentityUser? memberIdentity = await _memberManager.FindByIdAsync(model.UserId);
+        if (memberIdentity == null!)
+        {
+            return new ValidationErrorResult(
+                $"No local member found for user id {model.UserId}");
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.Website/Controllers/UmbLoginController.cs:154]: Similar to requesting a password reset, shouldn't reveal that the specific Id does not exist.

**Review by nul800sebastiaan — CHANGES_REQUESTED** [review]
Just had a look through the code so far and it's looking promising, I saw there was a question:

> However from the generated token we can't use this 

And then a link to `VerifyInvite` code. 
You state it can't be used, but not sure why not? 

Additionally I've left some notes around securely resetting passwords without revealing too much information to the public.

**Comment by bjarnef** [comment]
@nul800sebastiaan I think the `VerifyInvite` does something different for users:
https://github.com/umbraco/Umbraco-CMS/blob/96f8f70ed5a12765ceb3de1d174b492b70959df6/src/Umbraco.Web.BackOffice/Controllers/BackOfficeController.cs#L148

And I think we may want a different logic for members in frontend regarding password reset.

We can add logic to `UmbPasswordController` or maybe a `UmbMemberController`.

File: src/Umbraco.Web.Website/Controllers/UmbLoginController.cs (hunk 1 above)
Comment (bjarnef) [reply src/Umbraco.Web.Website/Controllers/UmbLoginController.cs:120]: I think I just copied this based on how it worked in backoffice users, when I think the `ValidationErrorResult` will show an error notification in backoffice.

Not sure if we should use this in frontend and return a different type.

**Comment by nul800sebastiaan** [comment]
@bjarnef Can you elaborate what needs to be different for members?

Yes `UmbPasswordController` would be where I'd go.

**Comment by bjarnef** [comment]
@nul800sebastiaan for members and frontend, we probably don't need the same logic, e.g. find a backoffice user here:
https://github.com/umbraco/Umbraco-CMS/blob/96f8f70ed5a12765ceb3de1d174b492b70959df6/src/Umbraco.Web.BackOffice/Controllers/BackOfficeController.cs#L185

and the redirect url after password reset probably need to be different as well.
https://github.com/umbraco/Umbraco-CMS/blob/96f8f70ed5a12765ceb3de1d174b492b70959df6/src/Umbraco.Web.BackOffice/Controllers/BackOfficeController.cs#L207

furthermore I am not sure what this line will do in frontend context:
https://github.com/umbraco/Umbraco-CMS/blob/96f8f70ed5a12765ceb3de1d174b492b70959df6/src/Umbraco.Web.BackOffice/Controllers/BackOfficeController.cs#L150

**Comment by nul800sebastiaan** [comment]
I don't have time to read all the code, but the approach should be the same for frontend and backoffice, yes, a few different redirect URLs but otherwise handling the token should be almost the same.


> furthermore I am not sure what this line will do in frontend context:  
Check the code for `AuthenticateBackOfficeAsync()`, we should probably have a `AuthenticateMemberAsync()` or something?

**Comment by bjarnef** [comment]
Yes, that's why I wanted to go with a specific `VerifyReset` instead although part of the code is the same :)

**Comment by nul800sebastiaan** [comment]
Aah okay, now I get your point! Yes, please go ahead, that sounds good! 👍

**Comment by bjarnef** [comment]
@nul800sebastiaan I wonder if there's a simply way to handle username/email in frontend based on `UsernameIsEmail` setting?
https://github.com/umbraco/Umbraco-CMS/blob/b923c3252537468a7ecad2d394f37c0dbc7387ae/src/Umbraco.Web.UI/appsettings.template.json#L46

but it seems `LoginModel` would just show `Username` in frontend rendering independent of this setting:
https://github.com/umbraco/Umbraco-CMS/blob/b923c3252537468a7ecad2d394f37c0dbc7387ae/src/Umbraco.Web.Common/Models/LoginModel.cs#L7-L9

**Comment by bjarnef** [comment]
@nul800sebastiaan I have a few more questions.

Regarding `AuthenticateMemberAsync()` which scheme should it use for member?

https://github.com/umbraco/Umbraco-CMS/pull/12991/files#diff-1416de4ad82b6443efbf013bb27c2895c1b6fd87610cdc2cc51ba94288b7847dR75-R82

In `SendMemberResetEmailAsync()` which area should it use at action variable?

**Comment by emmagarland** [comment]
Hi @bjarnef 

Just an update from HQ on this one

- The scheme should be `IdentityConstants.ApplicationScheme`.
- They don't think there is an area on the other snippets controllers like `UmbProfileController`, they should just be routing equally.

Hope that helps?

Thanks

Emma

**Comment by bjarnef** [comment]
@emmagarland I have updated the scheme.

Not totally sure about the `area` compared to user invite and since `UmbProfileController` doesn't use `_linkGenerator`:

```
// Get an mvc helper to get the URL
var action = _linkGenerator.GetPathByAction(
    nameof(UmbPasswordController.VerifyReset),
    ControllerExtensions.GetControllerName<UmbPasswordController>(),
    new { area = Core.Constants.Web.Mvc.BackOfficeApiArea, reset = resetToken });
```

Value of `Core.Constants.Web.Mvc.BackOfficeApiArea` used for user invite is "UmbracoApi" - not sure the same should be used for member password reset.

**Comment by emmagarland** [comment]
Hi @bjarnef 

Thanks for updating the scheme. 👍 

Just to let you know I have updated the Umbraco HQ team with your latest questions. I'll let you know any updates.

Best regards

Emma

**Comment by bergmania** [comment]
> Not totally sure about the `area` compared to user invite and since `UmbProfileController` doesn't use `_linkGenerator`:
> 
> ```
> // Get an mvc helper to get the URL
> var action = _linkGenerator.GetPathByAction(
>     nameof(UmbPasswordController.VerifyReset),
>     ControllerExtensions.GetControllerName<UmbPasswordController>(),
>     new { area = Core.Constants.Web.Mvc.BackOfficeApiArea, reset = resetToken });
> ```

I dont think you need an area at all. That's only required if the `UmbPasswordController` has an area. This method should just give you the correct link.

> The verify method for user invite call a `Default` method. We probably need something equivalent to member verify. https://github.com/umbraco/Umbraco-CMS/pull/12991/files#diff-5266e8c38c7620076446201ab1e96abb0a3d6bf5e91ffc0fa973cfa7186b9592R159-R190

I'm not sure about what you try to archive here. I guess the scenario is different per implementation. The default method of the user will show the backoffice when signed in or the login screen when not.

**Comment by bjarnef** [comment]
@bergmania okay, I will have another look without the area :) 

Regarding the `Default` method this part seems to handle redirect to backoffice and depending on runtime level.
https://github.com/umbraco/Umbraco-CMS/pull/12991/files#diff-5266e8c38c7620076446201ab1e96abb0a3d6bf5e91ffc0fa973cfa7186b9592R159-R190

and referenced to in `VerifyInvite()` - named it `VerifyReset()` in this context:
https://github.com/umbraco/Umbraco-CMS/pull/12991/files#diff-5266e8c38c7620076446201ab1e96abb0a3d6bf5e91ffc0fa973cfa7186b9592R192-R254

We probably don't need to worry much about the runtime level in this case.
Furthermore there is the `RedirectResult()` where it use a known format `#/login/false` although hardcoded, but regarding members it probably need to pass in a return URL, so unlike users who get an invite link to an alternative version of backoffice login page, members would be linked to a frontend page, which is different across projects.

**Comment by nul800sebastiaan** [comment]
Hey @bjarnef - I've been merging this and getting into this PR. It looks pretty good so far, but 2 things still stand out to me:

The email being sent includes these two sentences which make no sense for members:

> You have been invited by to the Umbraco Back Office.
> Message from :

image

I'd change those with something like: 

> Someone has requested a link to change your password. You can do this through the link below.
> [[ Button goes here ]]
> If you didn't request this, please ignore this email.


And the 2nd thing obviously is that clicking the link on the button goes nowhere at the moment. I would make sure that people can enter a URL where the "reset password" form will be hosted. I think I would pass like so: `ForgottenPasswordModel(string resetPasswordUrl)`. 
The URL is where people will be placing the reset password partial, which will need to read the querystring and make sure everything is good and it is a valid password reset request.

**Comment by bjarnef** [comment]
Hi @nul800sebastiaan 

Yes, the email was basically a copy if user invite email template. So we should update that. Maybe in the new backoffice the email template system could be better. Maybe sometime shared between CMS, Commerce, Forms, Workflow? I would like if all of the HTML wasn't part of the localization files - just the variables which vary by culture. Hopefully it will get better 😅

Regarding redirect my though was it could url the RedirectUrl property from `PostRedirectModel` which the model inherits from.. that way if can set this / adjust this in the partial view while still using default controller.

**Comment by nul800sebastiaan** [comment]
Yeah that could work too, so we keep it on the same page, but show / don't show either view output depending on if there's a reset querystring present. That could work quite nicely actually.

**Comment by bjarnef** [comment]
Yes, I think it was this, which is passed into the controller:
https://github.com/umbraco/Umbraco-CMS/pull/12991/files#diff-bef9cc04a94e6d59bf6a302585f9cf120b93c8dbd842209da5eeda74b029a58dR11

The login controller has something similar:
https://github.com/umbraco/Umbraco-CMS/pull/12991/files#diff-87235bb74b92194ed9c80cff50ea471d634c9a3a4840847641c962ed8458d28aR119

where login controller has fallback here if not specified:
https://github.com/umbraco/Umbraco-CMS/pull/12991/files#diff-87235bb74b92194ed9c80cff50ea471d634c9a3a4840847641c962ed8458d28aL65-L73

**Comment by bjarnef** [comment]
Regarding email templates there's a discussion about it here. I am not sure how much of this will change in v14, but at some point it can probably be enhanced and easier to manage. I guess CMS, Commerce (and Headless?) etc. could also benefits using a standardized for email templates. Some may be based on razor like in Umbraco Commerce and other just static HTML templates with dynamic variables?
https://github.com/umbraco/Umbraco-CMS/discussions/14631

**Comment by nul800sebastiaan** [comment]
Sure thing. But, as you know, email templating is not on the roadmap at the moment.

**Comment by iOvergaard** [comment]
@bjarnef There are no email templates in the new Backoffice - they all still reside on the server. Nor is there an email templating system planned. I am unsure how well this feature could work in the core - maybe a tutorial would be better suited if you need to configure a lot of things anyway to make it work. Do you still have an interest in updating this PR?

**Pull request closed without merging**